Resolve variables

langsamu opened the pull request:
Fixes #13 .

langsamu committed:
Resolve variables in step input fields

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+        step.processRow(meta, data);
+        step.processRow(meta, data);
+
+        File f = new File(expectedFilename);
```
Please mark all variables and parameters as `final` unless you are mutating them.

adamretter replied:
Within our code please use the newer `Path` API instead of `File`.
As many legacy APIs use File, you will find methods `File#toPath()` and `Path#toFile()` useful.

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+
+        File f = new File(expectedFilename);
+
+        assertTrue(f.exists() && !f.isDirectory());
```
e.g. `Files.exists(p)`

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    private StepDataInterface getData() {
+        return new JenaSerializerStepData() {{
```
Why `{{`? This is generally not a good idea... Better to create and then call `init()` please

langsamu replied:
Done.
Why?

adamretter replied:
So the following using the `{{` syntax:

```
new JenaSerializerStepData() {{ init(); }}
```

This actually creates a new anonymous sub-class of `JenaSerializerStepData` that inherits from `JenaSerializerStepData`. The constructor of the anonymous sub-class, will call `init()` on the anonymous sub-class, which effectively calls it on the super-class `JenaSerializerStepData`. Finally you create an instance of the anonymous sub-class with `new`.

Whereas the following code:
```
JenaSerializerStepData jssd = new JenaSerializerStepData();
jssd.init();
```

Simply constructs an instance of the class `JenaSerializerStepData`, and then calls its `init()` function.

----

1. There is no good reason that you need to create an anonymous sub-classes here. I understand you did it for code-brevity, but there is no technical reason for it, e.g. you don't need to add methods or override methods.
2. Also at debug time, the identity of an anonymous sub-class can be quite confusing.

p.s. You are only the second developer I have seen write `{{`. Most Java developers are not even aware of its existence or what it does. Many that I have explained it to think that it is a syntactic mistake ;-)

langsamu replied:
Perfect, thx.

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+
+    private static JenaSerializerStepMeta getMeta() {
+        return new JenaSerializerStepMeta() {{
+            setDefault();
```
Same comment as above.

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+
+        step.setRowHandler(
+                new StaticDataRowHandler(
+                        new ArrayList() {{
```
1. Please do not initialize objects in this way! `{{` should be avoided.

2. You are missing the generic typing specification.

Instead please use the longer:
```
final List<Object[]> list = new ArrayList<>();
list.add(...)
```

or instead:
```
final List<Object[]> list = Arrays.asList(...)
```

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    private class StaticDataRowHandler implements RowHandler {
+        private final LinkedList<Object[]> rows;
```
Do you really want a `LinkedList`?

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+            if (rows.isEmpty()) {
+                return null;
+            } else {
+                return rows.remove();
```
Is this meant to be a `pop`?

langsamu replied:
No. Proper type makes it clearer:
```java
private final Queue<Object[]> rows;
```

adamretter replied:
Okay. In Java if you don't need concurrent access to a queue, I would suggest instead using ArrayDeque and it's interface Deque.

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+import org.pentaho.di.trans.step.RowHandler;
+import org.pentaho.di.trans.step.StepDataInterface;
+import org.pentaho.di.trans.steps.mock.StepMockHelper;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
90% of the time, things should not be imported from the `sun` namespace as these are internal.

Rather than `NotImplementedException` for your purposes you should instead use `java.lang.UnuspportedOperationException`.

adamretter commented on src/test/java/uk/gov/nationalarchives/pdi/step/jena/serializer/JenaSerializerStepIT.java:
```diff
@@ -0,0 +1,192 @@
+
+    @Test
+    public void resolves_filename_variable() throws KettleException {
+        // TODO(SL): Use temp folder instead of random file
```
With Junit5 you can write:

```
@Test
public void resolves_filename_variable(@TempDir final Path tempDir) throws KettleException {
```

langsamu committed:
Assorted code improvements to #14

Make generic types specific.
Use java.nio (Files, Paths).
Explicit immutables.
Remove initializer blocks.
Refine typing.
Throw unsupported instead of not implemented.
Use tempdir.
Remove deprecated.

See review https://github.com/nationalarchives/kettle-jena-plugins/pull/14#pullrequestreview-579203017

langsamu committed:
Use appropriate queue interface and implementation

adamretter reviewed (approved):
LGTM, thanks for the changes :-)
